src/SSTClosGovernor.cpp

```diff
@@ -0,0 +1,196 @@
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:1:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_max);
+
+        // First low-priority bucket. Initially just give it the bottom of the
+        // turbo range. Potentially go lower at run time.
+        m_platform_io.write_control("SST::COREPRIORITY:2:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 0.67);
+        m_platform_io.write_control("SST::COREPRIORITY:2:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:2:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, (m_frequency_sticker + m_frequency_max) / 2);
+
+        // Least prioritized bucket. Initially just give it the bottom of the
+        // turbo range. Potentially go lower at run time.
+        m_platform_io.write_control("SST::COREPRIORITY:3:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 1.0);
+        m_platform_io.write_control("SST::COREPRIORITY:3:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:3:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+    }
+
+    int SSTClosGovernorImp::clos_domain_type(void) const
+    {
+        return m_clos_assoc_ctl_domain_type;
+    }
+
+    void SSTClosGovernorImp::adjust_platform(
+        const std::vector<double> &clos_by_core)
+    {
+        if (clos_by_core.size() != m_clos_control_idx.size()) {
+            throw Exception("SSTClosGovernorImp::" + std::string(__func__) +
+                                "(): size of request vector does not match size of control domain.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        m_do_write_batch = m_is_enabled && clos_by_core != m_last_clos;
+        if (m_is_enabled) {
+            for (size_t idx = 0; idx < m_clos_control_idx.size(); ++idx) {
+                m_platform_io.adjust(m_clos_control_idx[idx], clos_by_core[idx]);
+            }
+            m_last_clos = clos_by_core;
+        }
+    }
+
+    bool SSTClosGovernorImp::do_write_batch(void) const
+    {
+        return m_do_write_batch;
+    }
+
+    void SSTClosGovernorImp::enable_sst_turbo_prioritization()
+    {
+        // Enable prioritized turbo by cores
+        m_platform_io.write_control("SST::COREPRIORITY_ENABLE:ENABLE",
+                                    GEOPM_DOMAIN_BOARD, 0, 1);
+
+        // Enable the ability to extend the turbo range of high priority cores
+        // by decreasing the turbo range of low priority cores
+        m_platform_io.write_control("SST::TURBO_ENABLE:ENABLE",
+                                    GEOPM_DOMAIN_BOARD, 0, 1);
+
+        m_is_enabled = true;
```

Comment:
Why do we have `m_is_enabled = true;` at the end of `enable_sst_turbo_prioritization()` function, whereas the analogous statement it at the start of the `disable_sst_turbo_prioritization()` function? I think that we should probably be consistent for readability purposes, unless there is some kind of meaning of semantic difference between the enableing process and the disableing process. Is the property enabled only explicitly *after* performing those two `write_control()` calls?

Reply:
This class offers a high-level enable/disable abstraction over two dependencies' low-level enable/disable state. The high-level wrapper is only enabled while all of its low-level dependencies are enabled. The high-level wrapper is no longer fully enabled once we start disabling its dependencies.

If for some reason any error occurs part-way through enable/disable, the `m_is_enabled` state indicates whether the high-level wrapper is enabled.